docs/proposals/csi.md

```diff
@@ -0,0 +1,350 @@
+
+PersistentVolumeClaim Example:
+```yaml
+apiVersion: v1
+kind: PersistentVolumeClaim
+metadata:
+  name: csi-nfs-pvc
+spec:
+  accessModes:
+  - ReadWriteMany
+  resources:
+    requests:
+      storage: 1Gi
+  selector:
+    matchExpressions:
+    - key: name
+      operator: In
+      values: [" csi-nfs-pv"]
+```
+Based on the PersistentVolume, this example is used to define PersistentVolumeClaim.
+Once the PersistentVolumeClaim is created successfully,
+the PersistentVolume will be bound to the PersistentVolumeClaim.
+
+Pod binding PersistentVolumeClaim Example:
+```yaml
+kind: Pod
+apiVersion: v1
+metadata:
+  name: my-csi-app
+spec:
+  containers:
+    - name: my-frontend
+      image: busybox
+      volumeMounts:
+      - mountPath: "/data"
+        name: my-csi-volume
+      command: [ "sleep", "1000000" ]
+  volumes:
+    - name: my-csi-volume
+      persistentVolumeClaim:
+        claimName: csi-nfs-pvc
+```
+`claimName: csi-hostpath-pvc` is the name of PersistentVolumeClaim.
+Before the pod starts up, the volume will be attached to the edge node, and then mounted inside the pod.
+
+#### Dynamic Provisioning Example
+
+Dynamic Provisioning means the volume will be created by `External-Provisioner` firstly,
+and then the volume will be attached into the edge node which is running the pod,
+at last the volume will be mounted into the pod.
+Here list the CSI HostPath Driver as an example.
+
+StorageClass Example:
+```yaml
+apiVersion: storage.k8s.io/v1
+kind: StorageClass
+metadata:
+  name: csi-hostpath-sc
+provisioner: csi-hostpath-driver
+reclaimPolicy: Delete
+volumeBindingMode: Immediate
+```
+`provisioner: csi-hostpath-driver` is used to specify which kind of CSI Driver will be used.
+`csi-hostpath` is defined in the CSI hostPath Driver.
+
+PersistentVolumeClaim Example:
+```yaml
+apiVersion: v1
+kind: PersistentVolumeClaim
+metadata:
+  name: csi-hostpath-pvc
+spec:
+  accessModes:
+  - ReadWriteOnce
+  resources:
+    requests:
+      storage: 1Gi
+  storageClassName: csi-hostpath-sc
+```
+`storageClassName: csi-hostpath-sc` is used to specify the storage class name
+which is defined in the StorageClass Example.
+Once PersistentVolumeClaim is created, KubeEdge will create volume based on
+the CSI Driver which is specified in the StorageClass, and the PersistentVolume
+will be created in Kubernetes and bound to the PersistentVolumeClaim.
+
+Pod binding PersistentVolumeClaim Example:
+```yaml
+kind: Pod
+apiVersion: v1
+metadata:
+  name: my-csi-app
+spec:
+  containers:
+    - name: my-frontend
+      image: busybox
+      volumeMounts:
+      - mountPath: "/data"
+        name: my-csi-volume
+      command: [ "sleep", "1000000" ]
+  volumes:
+    - name: my-csi-volume
+      persistentVolumeClaim:
+        claimName: csi-hostpath-pvc
+```
+`claimName: csi-hostpath-pvc` is the name of PersistentVolumeClaim.
+Before the pod starts up, the volume will be attached to the edge node, and then mounted inside the pod.
+
+## Work Items
```

Comment:
Is there a need to store any of the pv/pvc metadata at the edge node in a db in case of offline scenarios ?

Reply:
yep. The pv/pvc metadata is more like the configmap and secret.
and the dataflow will over the metamanager.
currently we have key/value table to store the data.
Perhaps we can keep it:)
https://github.com/kubeedge/kubeedge/blob/master/edge/pkg/metamanager/msg_processor.go#L151